core/src/main/java/jenkins/install/SetupWizard.java

```diff
@@ -192,6 +208,155 @@ public void doCreateAdminUser(StaplerRequest req, StaplerResponse rsp) throws IO
         }
     }
 
+    /*package*/ static void completeUpgrade(Jenkins jenkins) throws IOException {
+        setCurrentLevel(Jenkins.getVersion());
+        // jenkins.getInstallState().proceedToNextState();
+    }
+    
+    /*package*/ static void setCurrentLevel(VersionNumber v) throws IOException {
+        FileUtils.writeStringToFile(getUpdateStateFile(), v.toString());
+    }
+    
+    /**
+     * File that captures the state of upgrade.
+     *
+     * This file records the version number that the installation has upgraded to.
+     */
+    /*package*/ static File getUpdateStateFile() {
+        return new File(Jenkins.getInstance().getRootDir(),"jenkins.install.UpgradeWizard.state");
+    }
+    
+    /**
+     * What is the version the upgrade wizard has run the last time and upgraded to?
+     */
+    public static VersionNumber getCurrentLevel() throws IOException {
+        VersionNumber from = new VersionNumber("1.0");
+        File state = getUpdateStateFile();
+        if (state.exists()) {
+            from = new VersionNumber(defaultIfBlank(readFileToString(state), "1.0").trim());
+        }
+        return from;
+    }
+    
+    /**
+     * Returns the initial plugin list in JSON format
+     */
+    @Restricted(DoNotUse.class) // WebOnly
+    public HttpResponse doPlatformPluginList() throws IOException {
+        jenkins.install.SetupWizard setupWizard = Jenkins.getInstance().getSetupWizard();
+        if (setupWizard != null) {
+            if (InstallState.UPGRADE.equals(Jenkins.getInstance().getInstallState())) {
+                JSONArray initialPluginData = getPlatformPluginUpdates();
+                if(initialPluginData != null) {
+                    return hudson.util.HttpResponses.okJSON(initialPluginData);
+                }
+            } else {
```

Comment:
Install States cannot be contributed by plugins in the current state AFAIK. I think it's in the scope of https://github.com/jenkinsci/jenkins/pull/2281